proof/crefine/RISCV64/Schedule_C.thy

```diff
@@ -593,35 +593,957 @@ lemma length_scRefills_bounded:
+  supply sched_context_C_size[simp del] refill_C_size[simp del]
+  unfolding scheduleUsed_def haskell_assert_def
+  apply (rule ccorres_symb_exec_l'[OF _ _ scActive_sp]; (solves wpsimp)?)
+  apply (rule ccorres_symb_exec_l'[OF _ _ assert_sp]; (solves wpsimp)?)
+  apply (cinit' lift: sc_' new_')
+   apply (rule_tac xf'="\<lambda>s. h_val (hrs_mem (t_hrs_' (globals s))) (ret__ptr_to_struct_refill_C_' s)"
+                in ccorres_split_nothrow_call)
+          apply (fastforce intro: refill_tail_ccorres, fastforce+)
+      apply ceqv
+     apply (rule ccorres_Guard_Seq)
+     apply csymbr
+     apply (rule ccorres_cond[where R=\<top>])
+       apply (clarsimp simp: crefill_relation_def)
+      apply (clarsimp simp: setRefillTl_def updateRefillTl_def)
+      apply (rule updateSchedContext_ccorres_lemma3
+                   [where P="\<lambda>sc. scRefillTail sc < length (scRefills sc)" and P'="?abs"])
+        apply vcg
+       apply fastforce
+      apply clarsimp
+      apply (rename_tac sc sc')
+      apply (frule (1) sc_ko_at_valid_objs_valid_sc'[OF _ invs_valid_objs'])
+      apply (frule rf_sr_refill_buffer_relation)
+      apply (frule_tac n="scRefillTail sc" in h_t_valid_refill, fastforce+)
+      apply (intro conjI)
+         apply (clarsimp simp: typ_heap_simps)
+        apply (rule disjCI2)
+        apply (rule_tac n="length (scRefills sc)" in array_assertion_shrink_right)
+         apply (fastforce intro: sc_at_array_assertion')
+        apply (clarsimp simp: typ_heap_simps csched_context_relation_def)
+       apply (clarsimp simp: sc_ptr_to_crefill_ptr_def typ_heap_simps csched_context_relation_def)
+      apply (clarsimp simp: crefill_relation_def csched_context_relation_def)
+      apply (frule rf_sr_refill_buffer_relation)
+      apply (frule h_t_valid_clift_Some_iff[THEN iffD1])
+      apply (clarsimp cong: StateSpace.state.fold_congs)
+      apply (rule_tac old_sc=sc and n="scRefillTail sc"
+                      and fa="rAmount_update (\<lambda>_. rAmount rv + rAmount new)"
+                       in rf_sr_refill_update2; fastforce?)
+         apply (fastforce simp: sched_context.expand)
+        apply (clarsimp simp: typ_heap_simps)
+        apply (fastforce simp: typ_heap_simps' sc_ptr_to_crefill_ptr_def crefill_relation_def)
+       apply (fastforce simp: csched_context_relation_def)
+      apply (fastforce dest: crefill_relationD
+                       simp: typ_heap_simps' sc_ptr_to_crefill_ptr_def crefill_relation_def)
+     apply (ctac add: refill_full_ccorres)
+       apply (rule ccorres_cond[where R=\<top>])
+         apply(clarsimp simp: to_bool_def)
+        apply (ctac add: refill_add_tail_ccorres)
+       apply (rule ccorres_rhs_assoc)
+       apply (rule ccorres_split_nothrow)
+           apply (clarsimp simp: setRefillTl_def updateRefillTl_def)
+           apply (rule updateSchedContext_ccorres_lemma3
+                        [where P="\<lambda>sc. scRefillTail sc < length (scRefills sc)" and P'="?abs"])
```

Comment:
`[` should be two spaces deeper than `u`, am seeing one on the github diff (saw this above as well in same file)

Reply:
Is there any particular reason that we would want two spaces here, rather than one? I thought that as long as it's strictly to the right of the name of the rule, it would be fine.

Reply:
as discussed offline: indent is nearly always by 2 spaces, except for subgoal-indentation where it's two spaces plus number of subgoals - 1
Also, there are other issues (the modification belongs to the lemma not to `rule`);
it might be better to have `..._lemma3[` to indicate the lemma is being modified, then wrap how it's modified onto the next line, with the `where` 2 in from `updateSched...`. Might be a tight fit still

Reply:
I think in these cases, it's nicest to just have it all on the one line, even though it does go over the 100 character blue line a little bit (by 5 characters or so). If that definitely doesn't sound good, then I can go with your suggestion.